drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/phases/TornadoTaskSpecialisation.java

```diff
@@ -271,29 +272,49 @@ protected void run(StructuredGraph graph, TornadoHighTierContext context) {
 
             deadCodeElimination.run(graph);
 
-            getDebugContext().dump(DebugContext.INFO_LEVEL, graph, "After TaskSpecialisation iteration=" + iterations);
+            getDebugContext().dump(DebugContext.INFO_LEVEL, graph, "After TaskSpecialisation iteration = " + iterations);
 
-            hasWork = (lastNodeCount != graph.getNodeCount() || graph.getNewNodes(mark).isNotEmpty()) // ||
-                                                                                                      // hasGuardingPiNodes)
-                    && (iterations < MAX_ITERATIONS);
+            hasWork = (lastNodeCount != graph.getNodeCount() || graph.getNewNodes(mark).isNotEmpty()) && (iterations < MAX_ITERATIONS);
             lastNodeCount = graph.getNodeCount();
             iterations++;
         }
 
+        graph.getNodes().filter(ParallelRangeNode.class).forEach(range -> {
```

Comment:
In cases of 2D parallel loops with both having `constants` as bounds. Does it change the bounds correctly or copies the outter bound to the inner range?

Reply:
+1 I will add a combination of those.